README.md
HashBuilder
===========
Used to build a hash result from contributed objects or hashes (presumably
properties on your object which should be considered in the isEqual: override).
The intention is for the hash result to be returned from an override to the
`NSObject` `- (NSUInteger)hash` method.

### Installing

If you're using [CocoPods](http://cocopods.org) it's as simple as adding this to your `Podfile`:

	pod 'HashBuilder', '~> 1.0'

### Documentation

To use, create a HashBuilder object, contribute to it, then query the 'builtHash'
property for the resulting hash.

	- (NSUInteger)hash
	{
		HashBuilder *builder = [HashBuilder builder];

		[builder contributeObject:self.objectID];
		[builder contributeObject:self.occurredDate];
		[builder contributeObject:self.type];
		[builder contributeObject:self.objectURL];
		[builder contributeObject:self.tags];
		[builder contributeObject:self.count];

		NSUInteger retVal = builder.builtHash;

		return retVal;
	}

It is prudent to consider the same properties when overriding your `- (BOOL)isEqual:(id)object`
method as well.

NOTE: The order of contribution _will_ change the resulting hash, even if all
the same values are contributed. For example:

    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:@"a"];
    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
    NSUInteger hash1 = builder1.builtHash;

    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder2 contributeObject:@"a"];
    NSUInteger hash2 = builder2.builtHash;

`hash1 != hash2`

#### Disclaimer and Licence

* This work makes use of techniques and concepts presented by [Mike Ash](http://www.mikeash.com/)
  in his post entitled [Implementing Equality and Hashing](http://www.mikeash.com/pyblog/friday-qa-2010-06-18-implementing-equality-and-hashing.html)
* This work is licensed under the [Creative Commons Attribution 3.0 Unported License](http://creativecommons.org/licenses/by/3.0/).
  Please see the included LICENSE.txt for complete details.

#### About

A professional iOS engineer by day, my name is Levi Brown. Authoring a technical
blog [grokin.gs](http://grokin.gs), I am reachable via:

Twitter [@levigroker](https://twitter.com/levigroker)  
App.net [@levigroker](https://alpha.app.net/levigroker)  
EMail [levigroker@gmail.com](mailto:levigroker@gmail.com)  

Your constructive comments and feedback are always welcome.


CHANGELOG.md
1.0   - February 12, 2013 - Initial release
1.0.1 - October 4, 2015   - Updated project and supporting files. No code changes.
1.0.2 - 2020/10/27 @levigroker - Minimum deployment version bump to silence Xcode 12 warnings.

Tests/HashBuilderTests.h
//
//  HashBuilderTests.h
//  HashBuilderTests
//
//  Created by Levi Brown on 2/12/13.
//  Copyright (c) 2013 Levi Brown. All rights reserved.
//

#import <SenTestingKit/SenTestingKit.h>

@interface HashBuilderTests : SenTestCase

@end


Tests/HashBuilderTests.m
//
//  HashBuilderTests.m
//  HashBuilderTests
//
//  Created by Levi Brown on 2/12/13.
//  Copyright (c) 2013 Levi Brown. All rights reserved.
//

#import "HashBuilderTests.h"
#import "HashBuilder.h"

@implementation HashBuilderTests

- (void)setUp
{
    [super setUp];
    
    // Set-up code here.
}

- (void)tearDown
{
    // Tear-down code here.
    
    [super tearDown];
}

- (void)test100
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:nil];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:nil];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test110
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:@"a"];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:@"a"];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test120
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:@"a"];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:@"b"];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test130
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test140
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:@"a"];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test150
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:@"a"];
    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:@"a"];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test160
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:@"a"];
    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder2 contributeObject:@"a"];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test170
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder1 contributeObject:@"a"];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder2 contributeObject:@"b"];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test180
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder1 contributeObject:@"a"];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder2 contributeObject:@"b"];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test190
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder1 contributeObject:@"a"];
    [builder1 contributeObject:@"b"];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder2 contributeObject:@"a"];
    [builder2 contributeObject:@"b"];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test200
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder1 contributeObject:@"a"];
    [builder1 contributeObject:@"b"];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:@"a"];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder2 contributeObject:@"b"];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test210
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
    [builder1 contributeObject:@"a"];
    [builder1 contributeObject:@"b"];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:@"b"];
    [builder2 contributeObject:@"a"];
    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test220
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:nil];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:@"a"];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test230
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:@"a"];
    [builder1 contributeObject:nil];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:@"a"];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test240
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:nil];
    [builder1 contributeObject:nil];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:nil];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test250
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:nil];
    [builder1 contributeObject:nil];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:nil];
    [builder2 contributeObject:nil];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test260
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeBOOL:YES];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeBOOL:YES];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test270
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeBOOL:YES];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeBOOL:NO];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test280
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeBOOL:YES];
    [builder1 contributeObject:nil];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeBOOL:YES];
    [builder2 contributeObject:nil];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test290
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeObject:nil];
    [builder1 contributeBOOL:YES];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:nil];
    [builder2 contributeBOOL:YES];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test300
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeBOOL:YES];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:nil];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test310
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeBOOL:NO];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeObject:nil];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test320
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeBOOL:NO];
    [builder1 contributeBOOL:NO];
    [builder1 contributeBOOL:NO];
    [builder1 contributeBOOL:NO];
    [builder1 contributeBOOL:NO];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeBOOL:NO];
    [builder2 contributeBOOL:NO];
    [builder2 contributeBOOL:NO];
    [builder2 contributeBOOL:NO];
    [builder2 contributeBOOL:NO];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test330
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeBOOL:YES];
    [builder1 contributeBOOL:YES];
    [builder1 contributeBOOL:YES];
    [builder1 contributeBOOL:YES];
    [builder1 contributeBOOL:YES];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeBOOL:YES];
    [builder2 contributeBOOL:YES];
    [builder2 contributeBOOL:YES];
    [builder2 contributeBOOL:YES];
    [builder2 contributeBOOL:YES];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertEquals(hash1, hash2, @"Hash '%lx' is unexpectedly not equal to '%lx'", hash1, hash2);
}

- (void)test340
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeBOOL:YES];
    [builder1 contributeBOOL:NO];
    [builder1 contributeBOOL:YES];
    [builder1 contributeBOOL:NO];
    [builder1 contributeBOOL:YES];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeBOOL:NO];
    [builder2 contributeBOOL:YES];
    [builder2 contributeBOOL:NO];
    [builder2 contributeBOOL:YES];
    [builder2 contributeBOOL:NO];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

- (void)test350
{
    HashBuilder *builder1 = [HashBuilder builder];
    [builder1 contributeBOOL:YES];
    [builder1 contributeBOOL:YES];
    [builder1 contributeBOOL:YES];
    [builder1 contributeBOOL:NO];
    [builder1 contributeBOOL:NO];
    NSUInteger hash1 = builder1.builtHash;
    
    HashBuilder *builder2 = [HashBuilder builder];
    [builder2 contributeBOOL:NO];
    [builder2 contributeBOOL:NO];
    [builder2 contributeBOOL:YES];
    [builder2 contributeBOOL:YES];
    [builder2 contributeBOOL:YES];
    NSUInteger hash2 = builder2.builtHash;
    
    STAssertTrue(hash1 != hash2, @"Hash '%lx' is unexpectedly equal to '%lx'", hash1, hash2);
}

@end


HashBuilder/HashBuilder.m
//
//  HashBuilder.m
//
//  Created by Levi Brown on 12/14/12.
//  Copyright (c) 2012, 2013 Levi Brown <mailto:levigroker@gmail.com>
//  This work is licensed under the Creative Commons Attribution 3.0
//  Unported License. To view a copy of this license, visit
//  http://creativecommons.org/licenses/by/3.0/ or send a letter to Creative
//  Commons, 444 Castro Street, Suite 900, Mountain View, California, 94041,
//  USA.
//
//  The above attribution and the included license must accompany any version
//  of the source code. Visible attribution in any binary distributable
//  including this work (or derivatives) is not required, but would be
//  appreciated.
//

#import "HashBuilder.h"

#define NSUINT_BIT (CHAR_BIT * sizeof(NSUInteger))
#define NSUINTROTATE(val, howmuch) ((((NSUInteger)val) << howmuch) | (((NSUInteger)val) >> (NSUINT_BIT - howmuch)))

@interface HashBuilder ()

@property (nonatomic,assign) NSUInteger contributionCount;
@property (nonatomic,assign) NSUInteger builtHash;

@end

@implementation HashBuilder

#pragma mark - Lifecycle

+ (id)builder
{
    id retVal = [[HashBuilder alloc] init];
    return retVal;
}

#pragma mark - Implementation

- (void)contributeBOOL:(BOOL)flag
{
    [self contributeHash:flag ? 1231 : 1237];
}

- (void)contributeObject:(id)obj
{
    [self contributeHash:[obj hash]];
}

- (void)contributeHash:(NSUInteger)hash
{
    if (hash == 0)
    {
        //Account for nil values as well.
        hash = 31;
    }
    
    if (self.builtHash == 0)
    {
        self.builtHash = hash;
        self.contributionCount = 1;
    }
    else
    {
        self.builtHash = NSUINTROTATE(hash, NSUINT_BIT / ++self.contributionCount) ^ self.builtHash;
    }
}

@end


HashBuilder/HashBuilder.h
//
//  HashBuilder.h
//
//  Created by Levi Brown on 12/14/12.
//  See: http://www.mikeash.com/pyblog/friday-qa-2010-06-18-implementing-equality-and-hashing.html
//  Copyright (c) 2012, 2013 Levi Brown <mailto:levigroker@gmail.com>
//  This work is licensed under the Creative Commons Attribution 3.0
//  Unported License. To view a copy of this license, visit
//  http://creativecommons.org/licenses/by/3.0/ or send a letter to Creative
//  Commons, 444 Castro Street, Suite 900, Mountain View, California, 94041,
//  USA.
//
//  The above attribution and the included license must accompany any version
//  of the source code. Visible attribution in any binary distributable
//  including this work (or derivatives) is not required, but would be
//  appreciated.
//
//

#import <Foundation/Foundation.h>

//Used to build a hash result from contributed objects or hashes (presumably properties
//on your object which should be considered in the isEqual: override).
//Create a HashBuilder object, contribute to it, then query the 'builtHash' property
//for the resulting hash.
//
//NOTE: The order of contribution /will/ change the resulting hash, even if all the same values are contributed.
//For example:
//
//    HashBuilder *builder1 = [HashBuilder builder];
//    [builder1 contributeObject:@"a"];
//    [builder1 contributeObject:[NSNumber numberWithInteger:12345]];
//    NSUInteger hash1 = builder1.builtHash;
//
//    HashBuilder *builder2 = [HashBuilder builder];
//    [builder2 contributeObject:[NSNumber numberWithInteger:12345]];
//    [builder2 contributeObject:@"a"];
//    NSUInteger hash2 = builder2.builtHash;
//
//hash1 != hash2

@interface HashBuilder : NSObject

@property (nonatomic,assign,readonly) NSUInteger builtHash;

+ (id)builder;

- (void)contributeBOOL:(BOOL)flag;
- (void)contributeObject:(id)obj;
- (void)contributeHash:(NSUInteger)hash;

@end
